src/v/coproc/tests/fixtures/fiber_mock_fixture.cc

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/tests/fixtures/fiber_mock_fixture.h"
+
+#include "coproc/api.h"
+#include "coproc/tests/utils/batch_utils.h"
+
+#include <seastar/core/coroutine.hh>
+
+namespace {
+
+class counting_consumer {
+public:
+    ss::future<ss::stop_iteration> operator()(const model::record_batch& rb) {
+        _num += rb.record_count();
+        co_return ss::stop_iteration::no;
+    }
+    std::size_t end_of_stream() { return _num; }
+
+private:
+    std::size_t _num{0};
+};
+
+} // namespace
+
+enable_coproc::enable_coproc() {
+    ss::smp::invoke_on_all([]() {
+        auto& config = config::shard_local_cfg();
+        config.get("coproc_offset_flush_interval_ms").set_value(500ms);
+        config.get("enable_coproc").set_value(true);
+    }).get0();
+}
+
+ss::future<> fiber_mock_fixture::init_test(test_parameters params) {
+    co_await _state.start();
+    co_await wait_for_controller_leadership();
+    co_await add_topic(params.tn, params.partitions);
+    for (auto i = 0; i < params.partitions; ++i) {
+        auto ntp = model::ntp(
+          params.tn.ns, params.tn.tp, model::partition_id(i));
+        auto shard = app.shard_table.local().shard_for(ntp);
+        vassert(shard, "topic not up yet");
+        co_await _state.invoke_on(
+          *shard, [this, ntp, params](state& s) -> ss::future<> {
+              auto src = co_await make_source(ntp, s, params);
+              auto [_, success] = s.routes.emplace(ntp, src);
+              vassert(success, "no double insert attempt should occur");
+          });
+    }
+}
+
+static ss::future<>
+do_verify_partition(storage::log log, model::ntp ntp, std::size_t n) {
+    auto cfg = storage::log_reader_config(
+      model::offset{0},
+      log.offsets().dirty_offset,
+      ss::default_priority_class());
+    cfg.type_filter = model::record_batch_type::raft_data;
+    auto rbr = co_await log.make_reader(cfg);
+    auto n_recs = co_await std::move(rbr).for_each_ref(
+      counting_consumer(), model::no_timeout);
+    BOOST_CHECK_EQUAL(n_recs, n);
+}
+
+ss::future<> fiber_mock_fixture::verify_results(expected_t expected) {
+    return _state.invoke_on_all(
+      [this, expected = std::move(expected)](state& s) mutable {
+          return ss::do_with(
+            std::move(expected), [this, &s](expected_t& expected) {
```

Comment:
I'm not sure this move is safe. I don't see a copy of `expected` per shard, perhaps that happens inside `invoke_on_all`?

Reply:
nice catch, I would expect this to fail to compile

Reply:
It's a bit tricky to follow, but it looks like the functor is moved into `invoke_on_all` and then forwarded through the call chain, so in general having that functor be `mutable` is probably a bad idea. I'd suggest moving `expected` into `do_with`, and then calling `invoke_on_all`, and removing all other mutations/moves of `expected`.

At least, I couldn't find any copies of functor anywhere, and the checks seem to be that it's nothrow movable.